# 융설구 09주차

## 데이터 수집 및 분석을 위한 파이썬 고급 문법

## 1. Python Comprehension
- 파이썬의 기본 자료구조는 알고 있다고 가정
    ```python
        list, dict, tuple, set
    ```
- 일반적인 `for`문 보다 빠릅니다.
- 가장 빈번하게 사용하는 `list comprehension`을 학습

### 1.1 List comprehension이 필요한 경우
- 1부터 10까지 각 수의 제곱을 담고 있는 리스트를 생성하라

In [1]:
a = []
for i in range(1, 10+1):
    a.append(i ** 2)

In [2]:
print(a)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


### 1부터 100까지 - 3의 배수만 $\to$ 리스트

In [3]:
b = []
for x in range(1, 101):
    if x % 3 == 0:
        b.append(x)

In [4]:
print(b)

[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99]


### 1.2 코드 한줄로 생성할 수 있는 방법 --> # List 내부에서 반복문을 돌리는 기능

In [5]:
# 리스트 내부에서 반복문을 돌리는 문법
a = [ i**2 for i in range(1, 11) ]

In [8]:
a

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

### 1.3 조건에 맞는 List를 생성할 수 있는 방법
- 예) 1부터 100까지 수 중에서 3의 배수를 포함하는 list를 생성하라...

In [6]:
# 조건을 이용해서 생성 가능
# 1 - 100 3의 배수만 담긴 리스트를 생성
a = [ x for x in range(1, 101) if x % 3 == 0]

In [9]:
a = [ i for i in range(1, 101) if i % 4 == 0]

In [10]:
print(a)

[4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100]


### 1.4 List 내부에 또 다른 Collection type을 편리하게 생성
- 다음과 같은 형태의 list를 만들어라.
- [ [1, 2, 3], [1, 2, 3], [1, 2, 3]]

In [11]:
a = []
for _ in range(3):
    a.append([ x for x in range(1, 4) ])

In [12]:
a

[[1, 2, 3], [1, 2, 3], [1, 2, 3]]

In [13]:
a = []
for _ in range(1, 4):
    elem = []
    for x in range(1, 4):
        elem.append(x)
    a.append(elem)

In [14]:
a

[[1, 2, 3], [1, 2, 3], [1, 2, 3]]

In [22]:
a = [ [ x for x in range(1, 4)] for _ in range(3)  ]

In [23]:
a

[[1, 2, 3], [1, 2, 3], [1, 2, 3]]

### 1.5 if-else를 결합한 List Comprehension
- 45보다 크거나 같으면 1을 더하고, 45보다 작으면 5를 더한 결과를 구하라.
- a = [22, 13, 45, 50, 98, 69, 43, 44, 1]
- 결과값 --> [27, 18, 46, 51, 99, 70, 48, 49, 6]

In [25]:
a = [22, 13, 45, 50, 98, 69, 43, 44, 1]

In [26]:
b = [ x+1 if x>=45 else x+45 for x in a]

In [27]:
b

[67, 58, 46, 51, 99, 70, 88, 89, 46]

In [10]:
b = [ x+1 if x >= 45 else x+5  for x in a  ]

### 1.6 성능비교

In [28]:
%%timeit
a = []
for x in range(1_000_000):
    a.append(x)
len(a)

56.5 ms ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [29]:
%%timeit
a = [ x for x in range(1_000_000) ]
len(a)

48.9 ms ± 742 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


### 1.7 Dictionary Comprehension

- 데이터를 사전 구조로 빠르게 생성하려는 경우
- 실습 데이터: week_09_dict_comprehension.xlsx

- shell 환경에서 pandas 설치
    
    ```bash
       $ sudo apt update
    ```
    
    ```bash
       sudo apt install python3-pip
    ```
    
- 주피터 노트북에서 설치

    ```bash 
       !sudo apt update && sudo apt upgrade -y
       !sudo apt install python3-pip
    ```
   
    
- 설치가 끝나면 노트북 서버를 다시 시작(현재 파일을 중지하고 커널 다시 시작)

In [30]:
!sudo apt update

Hit:1 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble InRelease
Get:2 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates InRelease [126 kB]
Get:3 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-backports InRelease [126 kB]
Get:4 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/main amd64 Packages [623 kB]
Get:5 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/main amd64 Components [114 kB]
Get:6 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/universe amd64 Packages [717 kB]
Get:7 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/universe Translation-en [213 kB]
Get:8 http://security.ubuntu.com/ubuntu noble-security InRelease [126 kB]      
Get:9 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/universe amd64 Components [305 kB]
Get:10 http://ap-northeast-2.ec2.archive.ubuntu.com/ubuntu noble-updates/restricted amd64 Components [212 B]
Get:11 http://ap-northeast-2.ec2

In [31]:
!sudo apt install python3-pip

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-pip is already the newest version (24.0+dfsg-1ubuntu1.1).
0 upgraded, 0 newly installed, 0 to remove and 17 not upgraded.


In [3]:
!pip install pandas

error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detai

```bash
$ pip3 install pandas
```

```bash
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
python3-xyz, where xyz is the package you are trying to
install.

If you wish to install a non-Debian-packaged Python package,
create a virtual environment using python3 -m venv path/to/venv.
Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
sure you have python3-full installed.

If you wish to install a non-Debian packaged Python application,
it may be easiest to use pipx install xyz, which will manage a
virtual environment for you. Make sure you have pipx installed.

See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or 
OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, 
by passing --break-system-packages.

hint: See PEP 668 for the detailed specification.
```


### 위와 같은 에러가 발생하는 이유 
- PEP [668](https://peps.python.org/pep-0668/) 참조
- StackOverflow [post](https://stackoverflow.com/questions/75602063/pip-install-r-requirements-txt-is-failing-this-environment-is-externally-mana)
- 해결방법
    - 전역(global)으로 pandas 패키지 설치 $\leftarrow$ 우리가 선택할 방법
        
        ```bash
        sudo apt install python3-pandas
        ```
    - 가상환경을 설치 $\to$ 가상환경에 jupyter notebook 설치 및 실행
    
        ```bash
        python -m venv venv
        . venv/bin/activate
        (venv) pip installl jupyter notebook
        (venv) python -m ipykernel install --user --name venv --display-name workspace-venv
        (venv) jupyter notebook
        ```

- 글로벌 패키지로 pandas 설치

In [32]:
!sudo apt install python3-pandas -y

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-pandas is already the newest version (2.1.4+dfsg-7).
0 upgraded, 0 newly installed, 0 to remove and 17 not upgraded.


In [33]:
import pandas as pd

### 만약 아래와 같은 메시지가 나온다면....

/home/ubuntu/.local/lib/python3.8/site-packages/pandas/core/computation/expressions.py:20: UserWarning: Pandas requires version '2.7.3' or newer of 'numexpr' (version '2.7.1' currently installed).


  from pandas.core.computation.check import NUMEXPR_INSTALLED

In [28]:
# !sudo apt install python3-numexpr -y

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 381.7/381.7 kB 17.2 MB/s eta 0:00:00
  Attempting uninstall: numexpr
    Found existing installation: numexpr 2.8.3
    Uninstalling numexpr-2.8.3:
      Successfully uninstalled numexpr-2.8.3


In [34]:
pd.__version__

'2.1.4'

In [38]:
# data = pd.read_csv( 파일명, 구분자, header=None  )
data = pd.read_csv('week_09_dict_comprehension.data', sep='\t', header=None)

In [39]:
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [40]:
df_data = pd.DataFrame(data)

In [6]:
df_data 

0   1
0   권나영  83
1   김민주  81
2   박건순  97
3   박민기  94
4   박상현  88
5   박성재  82
6   박지훈  87
7    박환  94
8   배수민  83
9   신민정  92
10  신윤찬  81
11  신재성  95
12  염상훈  81
13  오경준  85
14  원지연  98
15  유영길  84
16  유종엽  83
17  이선준  93
18  이우준  81
19  이은솔  83
20  이재구  95
21  이주엽  90
22  정혜원  96
23  지현구  81
24  차현욱  97
25  최준영  90
26  한석우  86
27  허선웅  89

In [7]:
df_data.columns

Index([0, 1], dtype='int64')

In [48]:
names = df_data[0].to_list()

In [49]:
values = df_data[1].tolist()

In [50]:
print(names)

['권나영', '김민주', '박건순', '박민기', '박상현', '박성재', '박지훈', '박환', '배수민', '신민정', '신윤찬', '신재성', '염상훈', '오경준', '원지연', '유영길', '유종엽', '이선준', '이우준', '이은솔', '이재구', '이주엽', '정혜원', '지현구', '차현욱', '최준영', '한석우', '허선웅']


In [51]:
print(values)

[83, 81, 97, 94, 88, 82, 87, 94, 83, 92, 81, 95, 81, 85, 98, 84, 83, 93, 81, 83, 95, 90, 96, 81, 97, 90, 86, 89]


### 1.7.1 names, values 사전으로 만들기 $\to$ 반복문 이용

In [52]:
dic_data = {} # 빈 사전 생성

In [53]:
for name, value in zip(names, values):
    dic_data[name] = value

In [54]:
print(dic_data)

{'권나영': 83, '김민주': 81, '박건순': 97, '박민기': 94, '박상현': 88, '박성재': 82, '박지훈': 87, '박환': 94, '배수민': 83, '신민정': 92, '신윤찬': 81, '신재성': 95, '염상훈': 81, '오경준': 85, '원지연': 98, '유영길': 84, '유종엽': 83, '이선준': 93, '이우준': 81, '이은솔': 83, '이재구': 95, '이주엽': 90, '정혜원': 96, '지현구': 81, '차현욱': 97, '최준영': 90, '한석우': 86, '허선웅': 89}


### 1.7.2 names, values 사전으로 만들기 $\to$ 좀 더 간편한 방법

In [55]:
dic_data2 = dict(zip(names, values))

In [56]:
print(dic_data2)

{'권나영': 83, '김민주': 81, '박건순': 97, '박민기': 94, '박상현': 88, '박성재': 82, '박지훈': 87, '박환': 94, '배수민': 83, '신민정': 92, '신윤찬': 81, '신재성': 95, '염상훈': 81, '오경준': 85, '원지연': 98, '유영길': 84, '유종엽': 83, '이선준': 93, '이우준': 81, '이은솔': 83, '이재구': 95, '이주엽': 90, '정혜원': 96, '지현구': 81, '차현욱': 97, '최준영': 90, '한석우': 86, '허선웅': 89}


### 1.7.3 names, values 사전으로 만들기 $\to$ Comphension 이용

In [59]:
dic_data3 = { name: value for name, value in zip(names, values)}

In [60]:
dic_data4 = {
    name: value   for name, value in zip(names, values)
}

In [61]:
print(dic_data3)

{'권나영': 83, '김민주': 81, '박건순': 97, '박민기': 94, '박상현': 88, '박성재': 82, '박지훈': 87, '박환': 94, '배수민': 83, '신민정': 92, '신윤찬': 81, '신재성': 95, '염상훈': 81, '오경준': 85, '원지연': 98, '유영길': 84, '유종엽': 83, '이선준': 93, '이우준': 81, '이은솔': 83, '이재구': 95, '이주엽': 90, '정혜원': 96, '지현구': 81, '차현욱': 97, '최준영': 90, '한석우': 86, '허선웅': 89}


## 퀴즈
- 위에서 만든 사전(Dictionary)의 Key와 Value를 바꾸기

In [62]:
dic_data_rev = { value: key for key, value in dic_data.items()  }

In [48]:
dic_data_rev = {
    y: x for x, y in dic_data.items()
}

In [63]:
print(dic_data_rev )

{83: '이은솔', 81: '지현구', 97: '차현욱', 94: '박환', 88: '박상현', 82: '박성재', 87: '박지훈', 92: '신민정', 95: '이재구', 85: '오경준', 98: '원지연', 84: '유영길', 93: '이선준', 90: '최준영', 96: '정혜원', 86: '한석우', 89: '허선웅'}


## 2. Callback 함수
- 함수를 argument parameter로 사용

In [64]:
def calculator(f, num1, num2):
    return f(num1, num2)

In [65]:
def plus(a, b):
    return a + b

In [66]:
def minus(a, b):
    return a - b

In [67]:
calculator(plus, 3, 4)

7

In [68]:
calculator(minus, 3, 4)

-1

In [69]:
def calculator2(indx, a, b):
    if indx == '+':
        return a + b
    elif indx == '1':
        return a - b
    elif indx == 'sq':
        return 'something'
    
    

In [70]:
calculator2('+',  2, 3)

5

In [71]:
def square(a, b):
    return a ** b

In [72]:
calculator(square, 5, 6)

15625

## 퀴즈
- 기존 함수: 구구단 출력 프로그램 작성
    - 함수 1: gugu_asc, 파라미터: dan <- 출력을 원하는 단
    - 함수 2: gugu_dsc, 파라미터: dan <- 출력을 원하는 단
- 기존 함수 2개를 이용하여 구구단 출력
    - 함수명: gugu_dir

In [73]:
def gugu_asc(dan):
    for x in range(1, 10):
        print(f'{dan} x {x} = {dan * x}')

In [75]:
def gugu_dsc(dan):
    for x in range(1, 10):
        print(f'{dan} x {10 - x} = {dan * (10 - x)}')

In [76]:
gugu_dsc(4)

4 x 9 = 36
4 x 8 = 32
4 x 7 = 28
4 x 6 = 24
4 x 5 = 20
4 x 4 = 16
4 x 3 = 12
4 x 2 = 8
4 x 1 = 4


In [77]:
def gugu_dir(f, dan):
    return f(dan)

In [79]:
gugu_dir(gugu_dsc, 8)

8 x 9 = 72
8 x 8 = 64
8 x 7 = 56
8 x 6 = 48
8 x 5 = 40
8 x 4 = 32
8 x 3 = 24
8 x 2 = 16
8 x 1 = 8


<hr>

## 3. Packing, Unpacking
- 파이썬 오픈 소스 등에서 매주 빈번히 보게되는 문법
- 내가 쓰지 않더라도 남이 작성한 코드는 이해할 줄 알아야 한다.
- 함수 파라미터 관리를 매우 유연하게 해줄 수 있는 문법

- 문법:
    - 파라미터 전달(list, tuple)
        - Packing: *
        - Unpacking: *
    - 키워드 파라미터 전달(dict)
        - Packing: **
        - Unpacking: **

### 3.1 Packing

- 만약 입력 파라미터의 개수가 3개보다 많거나 적다면?

In [80]:
# 다음 함수의 문제점?
#    함수의 기능: 입력 파라미터의 총합
def total_sum(a, b, c):
    return a + b + c

In [81]:
total_sum(1, 2, 3)

6

In [82]:
total_sum(a=1, b=2, c=3)

6

- 만약, 입력 파라미터 개수가 3개보다 많거나 적다면?

In [85]:
a = [1, 2, 3, 10, 15, 17]

In [86]:
def total_sum2(a):
    temp = 0
    for x in a:
        temp += x
    return temp

In [88]:
total_sum2(a)

48

In [90]:
total_sum2(1, 2, 3, 10, 15, 17)

TypeError: total_sum2() takes 1 positional argument but 6 were given

In [91]:
total_sum(1, 2)

TypeError: total_sum() missing 1 required positional argument: 'c'

- 해결책
    - tuple packing 기능 활용
    - 파라미터 수에 관계없이 작동하도록 변경

In [92]:
def total_sum2(*args):
    sum = 0
    for x in args:
        sum += x
    return sum

In [93]:
total_sum2(1)

1

In [94]:
a = 1, 2, 4, 19, 32, 23, 88

In [95]:
type(a)

tuple

In [96]:
total_sum2(*a)

169

- 파라미터 전달 및 패킹의 실체
    - 함수를 선언할 때는 패킹

In [99]:
a = 1, 2

In [100]:
x, y = a

In [101]:
x

1

In [102]:
y

2

In [104]:
def total_sum3(*args):
    print(type(args))
    sum = 0
    for x in args:
        sum += x
    return sum

In [105]:
total_sum3(1, 2, 4, 19, 32, 23, 88)

<class 'tuple'>


169

- Dict Packing의 활용

In [106]:
def family_name(mom, dad, sister, brother):
    print('엄마 이름은 {}입니다.'.format(mom))
    print('아빠 이름은 {}입니다.'.format(dad))
    print('누나 이름은 {}입니다.'.format(sister))
    print('형 이름은 {}입니다.'.format(brother))

In [107]:
family_name(mom='유관선', dad='홍길동', sister='홍이뻐', brother='노기섭')

엄마 이름은 유관선입니다.
아빠 이름은 홍길동입니다.
누나 이름은 홍이뻐입니다.
형 이름은 노기섭입니다.


- 만약 형이 있다면?

In [110]:
family_name(mom='유관선', dad='홍길동', sister='홍이뻐', brother='홍영웅', bro='큰형')

TypeError: family_name() got an unexpected keyword argument 'bro'

In [111]:
family_name(mom='유관선', dad='홍길동')

TypeError: family_name() missing 2 required positional arguments: 'sister' and 'brother'

- 해결책
    - 어떤 키워드 파라미터가 들어오더라도 처리할 수 있도록 변경
    - Dict packing 적용

In [112]:
def family_name2(**kwargs):
    for key, value in kwargs.items():
        print('{} 이름은 {}입니다.'.format(key, value))

In [113]:
family_name2(mom='유관선', dad='홍길동')

mom 이름은 유관선입니다.
dad 이름은 홍길동입니다.


In [114]:
family_name2(mom='유관선', dad='홍길동', sister='홍이뻐', bro='홍영웅', cat='나비')

mom 이름은 유관선입니다.
dad 이름은 홍길동입니다.
sister 이름은 홍이뻐입니다.
bro 이름은 홍영웅입니다.
cat 이름은 나비입니다.


- Dict 파라미터 전달 및 패킹의 실체

In [115]:
def family_name3(**kwargs):
    print(type(kwargs))
    for key, value in kwargs.items():
        print('{} 이름은 {}입니다.'.format(key, value))

In [116]:
family_name3(mom='유관선', dad='홍길동')

<class 'dict'>
mom 이름은 유관선입니다.
dad 이름은 홍길동입니다.


### 3.2 Unpacking

In [95]:
a = [1, 4, 6, 7] # list
b = (1, 4, 6, 7) # tuple

In [96]:
total_sum3(1, 4, 6, 7 )

<class 'tuple'>


18

In [97]:
total_sum3(*a)

<class 'tuple'>


18

In [98]:
a = 1, 2

In [99]:
x, y = a

In [100]:
x

1

In [101]:
y

2

In [102]:
total_sum3(1, 4, 6, 7)

<class 'tuple'>


18

In [103]:
total_sum3(*b)

<class 'tuple'>


18

In [117]:
my_family = {
    '아빠': '홍길동',
    '엄마': '이영희',
    '동생': '홍철기',
    '강아지': '바둑이',
}

In [118]:
def family_name3(**kwargs):
    print(type(kwargs))
    for key, value in kwargs.items():
        print('{} 이름은 {}입니다.'.format(key, value))    

In [119]:
family_name3(**my_family)

<class 'dict'>
아빠 이름은 홍길동입니다.
엄마 이름은 이영희입니다.
동생 이름은 홍철기입니다.
강아지 이름은 바둑이입니다.


## 4. Lambda Function
- 익명함수 (함수의 이름이 없는 함수)
- 함수를 3항 연산자와 비슷하게 한줄로 간결하게 표현 가능
- 생각보다 많이 활용되는 유용한 함수
- 1회용 함수로 자주 사용 (함수 호출시 생성, 실행 후 소멸) --> 메모리의 효율적 사용
- 간단한 함수에 주로 적용

In [120]:
def (*args, **kwargs):
    print(type(kwargs))
    for key, value in kwargs.items():
        print('{} 이름은 {}입니다.'.format(key, value))    

SyntaxError: invalid syntax (3773330007.py, line 1)

In [121]:
def plus(a, b):
    return a + b

In [122]:
plus(4, 5)

9

In [110]:
plus2 = lambda a, b: a + b

In [111]:
plus2(3, 4)

7

In [123]:
calculator(lambda a, b: a**2 + b**3, 5, 6)

241

## 5. map 함수
### List 원소들에게 특정 효과(또는 operation)을 적용하고자 할 경우
### 사용법: map(함수이름, 이터러블)

In [128]:
# 리스트 값 중에서 짝수, 홀수를 구분해 내는 기능을 만들고 싶어요
a = [1, 2, 3, 4]
def odd_or_even(num):
    return '짝수' if num % 2 == 0 else '홀수'

In [129]:
odd_or_even(5)

'홀수'

In [130]:
def abc(num_list):
    temp = []
    for x in num_list:
        temp.append(odd_or_even(x))
    return temp    

In [131]:
# List comprehension을 사용할 경우
def abc2(num_list):
    return [odd_or_even(x) for x in num_list]


In [132]:
abc(a)

['홀수', '짝수', '홀수', '짝수']

In [133]:
abc2(a)

['홀수', '짝수', '홀수', '짝수']

In [134]:
map(odd_or_even, a)

In [135]:
result = list( map(odd_or_even, a))

In [136]:
result

['홀수', '짝수', '홀수', '짝수']

In [121]:
result2 = map(lambda x: '짝수' if x % 2 == 0 else '홀수', a)

In [122]:
list(result2)

['홀수', '짝수', '홀수', '짝수']

## 퀴즈
- map 함수를 이용할 것!
- 사용자로부터 입력받은 데이터를 띄어쓰기를 기준으로 분할하여 리스트로 작성
- 리스트에 담기는 데이터는 정수형 숫자로 변환

In [137]:
data =  input('Type numbers: ')

Type numbers: 1 2 3


In [138]:
data

'1 2 3'

In [128]:
data

'1 2 3 4 5'

In [129]:
data2 = data.split(' ')
data2

['1', '2', '3', '4', '5']

In [130]:
result = [ int(x) for x in data2]

In [131]:
result

[1, 2, 3, 4, 5]

In [132]:
#result2 = map(함수이름, 이터러블)
result2 = map(lambda x : int(x), data.split(' '))

In [133]:
list(result2)

[1, 2, 3, 4, 5]

## 6. filter 함수
### 리스트 원소 중에서 조건에 맞는 원소만 남기고 싶을 때
### 사용법: filter(함수명, 이터러블)

In [139]:
a = list(range(20))

In [140]:
a

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]

In [137]:
def screen(x):
    if x % 2 == 0 or x % 3 == 0:
        return x

In [138]:
result = list(filter(screen, a))

In [139]:
result

[2, 3, 4, 6, 8, 9, 10, 12, 14, 15, 16, 18]

In [140]:
result = list(filter( lambda x : x % 2 == 0 or x % 3 == 0,   a))

In [141]:
result

[0, 2, 3, 4, 6, 8, 9, 10, 12, 14, 15, 16, 18]

## 7. reduce 함수 (수업 시간에는 생략)
### 리스트 요소를 반복적으로 연산을 적용하여 하나의 값으로 도출할 때
### reduce 함수를 임포트 하여 사용
    - from functools import reduce
### 사용법: reduce(함수이름, 이터러블)

In [141]:
from functools import reduce

In [142]:
a =  [5, 2, 3, 4]

In [143]:
def redu(x, y):
    return x - y

In [144]:
result3 = reduce(redu,  a)

In [145]:
result3

-4

In [146]:
result4 = reduce(lambda x, y: x - y,  a)
result4

-4

## 7. Decorators
- 함수를 받아 명령을 추가한 뒤 이를 다시 함수의 형태로 반환하는 함수
- 함수의 내부를 수정하지 않고 기능에 변화를 주고 싶을 때 사용
- 일반적으로 함수의 전처리나 후처리에 대한 필요가 있을때 사용
- 데코레이터를 이용해, 반복을 줄이고 메소드나 함수의 책임을 확장


- 즉, 함수에서 코드를 바꾸지 않고 기능을 추가하거나 수정할 때


### 7.1 코드가 중복되는 상황
```
def sample_func_1():
    soucce_code_1
    soucce_code_2
    soucce_code_3
    return 
```

```
def sample_func_2():
    soucce_code_1
    soucce_code_4
    soucce_code_3
    return 
```

- 코드의 중복
```soucce_code_1```과 ```soucce_code_3``` 각각의 함수에서 중복
- 만약, 여러개의 함수에서 중복되는 소스코드가 100군데라면???
- 중복되는 코드를 묶어서 따로 관리할 수 없을까??

### 7.2 Decorator 일반적 활용 (conventions)

- 일반적인 Decorator 선언 방법
```
def deco_func(func_name):

    # Decorator가 실행된 즉시 wrapper 함수가 선언됨
    def wrapper(*args, **kwargs):
        source_code
            :
            :
        
        result = func_name(*args, **kwargs)
        
        source_code
            :
            :
        
        return result # wrapper 함수의 리턴값
    
    return wrapper 
```



- Decorator 적용

```
@deco_func
def test1():
    source_code
        :
        :
```
- deco_func 함수가 리턴한 wrapper 함수가 test1 함수로 바뀜



```
@deco_func
def test2():
    source_code
        :
        :
```
- deco_func 함수가 리턴한 wrapper 함수가 test2 함수로 바뀜

## 7.3 Decorator 실습 - 함수의 실행 시간을 측정하는 데코레이터 작성

In [147]:
import time

### 데코레이터 함수 정의

In [150]:
def run_time_decorator(func_name):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func_name(*args, **kwargs)
        end_time = time.time()
        print('함수 실행시간: {}'.format(end_time - start_time))
        return result
    return wrapper        

### 일반 함수 정의

- 아래 계산 결과를 모두 더하는 함수 $\to$ big_loop(num)

\begin{align}
1^2 \\
1^2 + 2^2 \\
1^2 + 2^2 + 3^2 \\
   : \\
1^2 + 2^2 + 3^2 + .... + num^2 \\
\end{align}




In [149]:
def big_loop(num):
    data = list(range(1, num+1))
    result = []
    for x in data:
        temp = 0
        for y in range(1, x+1):
            temp += y ** 2
        result.append(temp)
    print('결과값은 {} 입니다.'.format(sum(result)))
    return sum(result)

### 데코레이터를 사용하지 않는 경우

In [151]:
start_time = time.time()
big_loop(1000)
end_time = time.time()
print('함수 실행시간: {}'.format(end_time - start_time))

결과값은 83667083500 입니다.
함수 실행시간: 0.05041766166687012


### 데코레이터를 사용할 경우

In [154]:
@run_time_decorator
def big_loop2(num):
    data = list(range(1, num+1))
    result = []
    for x in data:
        temp = 0
        for y in range(1, x+1):
            temp += y ** 2
        result.append(temp)
    print('결과값은 {} 입니다.'.format(sum(result)))
    return sum(result)

In [155]:
big_loop2(1000)

결과값은 83667083500 입니다.
함수 실행시간: 0.05078315734863281


83667083500

In [157]:
@run_time_decorator
def big_loop3(num):
    data = list(range(1, num+1))
    result = []
    for x in data:
        temp = 0
        for y in range(1, x+1):
            temp += y ** 2
        result.append(temp)
    print('결과값은 {} 입니다.'.format(sum(result)))
    return sum(result)

In [158]:
big_loop3(1000)

결과값은 83667083500 입니다.
함수 실행시간: 0.04899907112121582


83667083500

### 또 다른 일반 함수로 실습
- 배열의 제곱 합을 구하는 함수

In [159]:
def list_sum(num):
    data = range(1, num+1)
    result = 0
    for x in data:
        result += x ** 2
    print('결과값: {}'.format(result))
    return result

In [160]:
list_sum(1000)

결과값: 333833500


333833500

### 만약 시간측정 데코레이터가 있다면 간단히 해결!

In [161]:
@run_time_decorator
def list_sum2(num):
    data = range(1, num+1)
    result = 0
    for x in data:
        result += x ** 2
    print('결과값: {}'.format(result))
    return result

In [162]:
list_sum2(1000)

결과값: 333833500
함수 실행시간: 0.0006856918334960938


333833500

### Notebook Magic Command 사용해 보기

In [167]:
%%timeit
def list_sum2(num):
    data = range(1, num+1)
    result = 0
    for x in data:
        result += x ** 2
    print('결과값: {}'.format(result))
    return result

60.8 ns ± 0.24 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


## 퀴즈
- 로그인 함수: log_in(user_name, password), 리턴: True / Fasle
- 임의의 함수: print_hello(name) $\to$ 'Hello {name}'를 출력하는 함수
- 데코레이터 함수: login_required
    - user_name 과 password 가 일치하면 print_hello 함수 실행
    - user_name 과 password 가 틀리면 '로그인이 필요한 서비스 입니다.' 출력

In [163]:
LOGIN = True
#LOGIN = False

In [164]:
def print_hello(name):
    print(f'Hello {name}')

In [165]:
# 데코레이터가 없는 경우

if LOGIN:
    print_hello("교수님")
else:
    print('로그인이 필요한 서비스 입니다.')

Hello 교수님


<hr>

In [166]:
# 데코레이터 함수
def login_required(func_name):
    def wrapper(*args, **kwargs):
        if LOGIN:
            return func_name(*args, **kwargs)
        else:
            print('로그인이 필요한 서비스 입니다.')
            return
            
    return wrapper     

In [168]:
@login_required
def print_hello(name):
    print(f'Hello {name}')

In [169]:
print_hello('교수님')

Hello 교수님


In [170]:
# LOGIN = True
LOGIN = False

In [171]:
print_hello('홍길동')

로그인이 필요한 서비스 입니다.
